Use native file I/O in local and overlay drive

weirddan455 opened the pull request:
# Description

This PR replaces the use of C library I/O (`fopen()` and friends) with native file I/O.  Windows uses Win32 API calls and everyone else uses POSIX.

I was also able to remove some hacky `ftell()` + `fseek()` nonsense that should not be needed since there's no library buffering.  The OS itself will still buffer but that should be fine.  All we care about is that a write followed by a read will return the new data.

I have some more work on the file I/O planned to better handle timestamps (which will depend on native file I/O) but I thought it would be best to put that in a separate PR and get this in first.

# Manual testing

Tested on Windows (both MSYS2/GCC and MSVC/Clang) and Linux.  Crystal Caves and Mortal Kombat saves work which had been problematic on Windows in the past.

Other games with save state tested:  Secret of Monkey Island 2, Quake, Doom, Space Quest 1

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

johnnovak commented on include/dos_system.h:
```diff
@@ -189,7 +190,7 @@ class DOS_Device : public DOS_File {
 class localFile : public DOS_File {
 public:
-	localFile(const char* name, const std_fs::path& path, FILE* handle,
+	localFile(const char* name, const std_fs::path& path, NativeFileHandle handle,
```
`const NativeFileHandle`?

weirddan455 replied:
fixed

johnnovak commented on include/fs_utils.h:
```diff
@@ -32,6 +32,37 @@
+#else
+using NativeFileHandle = int;
+constexpr NativeFileHandle InvalidNativeFileHandle = -1;
+#endif
```
Maybe format it like this to make it look marginally nicer:

```cpp
#if defined(WIN32)
	#include <windows.h>

	using NativeFileHandle = HANDLE;
	// Cannot be constexpr due to Win32 macro
	#define InvalidNativeFileHandle INVALID_HANDLE_VALUE

#else // Linux, macOS
	using NativeFileHandle = int;
	constexpr NativeFileHandle InvalidNativeFileHandle = -1;
#endif
```

weirddan455 replied:
fixed

johnnovak commented on include/fs_utils.h:
```diff
@@ -32,6 +32,37 @@
+{
+	uint16_t date = 0;
+	uint16_t time = 0;
+};
```
Just run the formatter on these. Then you'll get stuff like this:

```cpp
enum class NativeFileSeek { Set, Current, End };
```

weirddan455 replied:
formatted

johnnovak commented on include/fs_utils.h:
```diff
@@ -168,12 +199,31 @@ std::deque<std_fs::path> get_xdg_data_dirs() noexcept;
                                     const FatAttributeFlags attributes);
 
+// Windows ReadFile and WriteFile take in a DWORD (32 bit unsigned) for number of bytes
+// We'll make that our limit until someone can give a good reason why they need to read/write more than 4GB (unlikely)
```
Indeed 😎

FeralChild64 replied:
@johnnovak And the maximum file size for FAT32 is still 4GB.

weirddan455 replied:
Removed this comment.  It can handle large files now.

johnnovak commented on include/fs_utils.h:
```diff
@@ -168,12 +199,31 @@ std::deque<std_fs::path> get_xdg_data_dirs() noexcept;
+// Windows ReadFile and WriteFile take in a DWORD (32 bit unsigned) for number of bytes
+// We'll make that our limit until someone can give a good reason why they need to read/write more than 4GB (unlikely)
+NativeFileReturn local_drive_read_file(const NativeFileHandle handle, uint8_t *buffer, const uint32_t bytes_requested);
+NativeFileReturn local_drive_write_file(const NativeFileHandle handle, const uint8_t *buffer, const uint32_t bytes_requested);
```
I guess we could even just use `int` for all these byte sizes because of the FAT16 limits if it's easier overall.

weirddan455 replied:
I suppose I can go up to `int64_t`.  I would need to write some additional code to break up large (greater than 4GB) reads/writes on Windows into multiple calls but that'll be easy enough.

weirddan455 replied:
Changed to `int64_t`

johnnovak commented on include/fs_utils.h:
```diff
@@ -168,12 +199,31 @@ std::deque<std_fs::path> get_xdg_data_dirs() noexcept;
+NativeFileReturn local_drive_read_file(const NativeFileHandle handle, uint8_t *buffer, const uint32_t bytes_requested);
+NativeFileReturn local_drive_write_file(const NativeFileHandle handle, const uint8_t *buffer, const uint32_t bytes_requested);
+
+int32_t local_drive_seek_file(const NativeFileHandle handle, const int32_t offset, const NativeFileSeek type);
```
Just `int offset`? Or `int64`?

weirddan455 replied:
Changed to `int64_t`

johnnovak commented on include/fs_utils.h:
```diff
@@ -168,12 +199,31 @@ std::deque<std_fs::path> get_xdg_data_dirs() noexcept;
+int32_t local_drive_seek_file(const NativeFileHandle handle, const int32_t offset, const NativeFileSeek type);
+
+// Returns offset from beginning of file
+int32_t local_drive_get_file_position(const NativeFileHandle handle);
```
`int64`?

weirddan455 replied:
Changed to `int64_t`

johnnovak commented on include/fs_utils.h:
```diff
@@ -32,6 +32,37 @@
+
+struct NativeFileReturn
+{
+	uint32_t bytes = 0;
```
`int64`?

johnnovak replied:
`num_bytes` would be better.

weirddan455 replied:
Changed to `int64_t` and renamed

johnnovak commented on src/misc/fs_utils_posix.cpp:
```diff
@@ -354,4 +357,104 @@ uint16_t local_drive_set_attributes(const std_fs::path& path,
+// POSIX does not guarantee to read or write all bytes requested at once.
+// On Linux, it will read the entire chunk in one call (assuming it's a normal file).
+// On Windows, it does not always do so (although we're not using these functions on Windows).
+// On Mac and the various BSDs, who knows. So be safe and do a loop.
```
Nice explanation 👍🏻 And yeah, this is the correct way to read a file on all platforms. Even Linux might change things up tomorrow.

weirddan455 replied:
Linux is pretty strict about not breaking userspace so I doubt this behavior will change.  I suppose someone could mount their DOSBox drive on a networked filesystem or something though and then I'm not sure what the behavior is but the loop should work everywhere.

johnnovak commented on src/misc/fs_utils_posix.cpp:
```diff
@@ -354,4 +357,104 @@ uint16_t local_drive_set_attributes(const std_fs::path& path,
+	ret.bytes = 0;
+	ret.error = false;
+	while (ret.bytes < bytes_requested) {
+		const auto bytes = read(handle, buffer + ret.bytes, bytes_requested - ret.bytes);
```
`num_bytes_read`?

weirddan455 replied:
renamed

johnnovak commented on src/misc/fs_utils_posix.cpp:
```diff
@@ -354,4 +357,104 @@ uint16_t local_drive_set_attributes(const std_fs::path& path,
+// On Linux, it will read the entire chunk in one call (assuming it's a normal file).
+// On Windows, it does not always do so (although we're not using these functions on Windows).
+// On Mac and the various BSDs, who knows. So be safe and do a loop.
+NativeFileReturn local_drive_read_file(const NativeFileHandle handle, uint8_t *buffer, const uint32_t bytes_requested)
```
`num_bytes_requested`?

weirddan455 replied:
renamed

johnnovak commented on src/misc/fs_utils_posix.cpp:
```diff
@@ -354,4 +357,104 @@ uint16_t local_drive_set_attributes(const std_fs::path& path,
+	return ret;
+}
+
+NativeFileReturn local_drive_write_file(const NativeFileHandle handle, const uint8_t *buffer, const uint32_t bytes_requested)
```
`num_bytes_requested`?

weirddan455 replied:
renamed

johnnovak commented on src/misc/fs_utils_posix.cpp:
```diff
@@ -354,4 +357,104 @@ uint16_t local_drive_set_attributes(const std_fs::path& path,
+	ret.bytes = 0;
+	ret.error = false;
+	while (ret.bytes < bytes_requested) {
+		const auto bytes = write(handle, buffer + ret.bytes, bytes_requested - ret.bytes);
```
`num_bytes_written`?

weirddan455 replied:
renamed

johnnovak commented on src/misc/fs_utils_posix.cpp:
```diff
@@ -354,4 +357,104 @@ uint16_t local_drive_set_attributes(const std_fs::path& path,
+int32_t local_drive_seek_file(const NativeFileHandle handle, const int32_t offset, const NativeFileSeek type)
+{
+	int posix_seek_type = SEEK_SET;
+	switch(type) {
```
Space after `switch`.

johnnovak replied:
Well, just reformat the whole file; formatting is off.

weirddan455 replied:
formatted

johnnovak commented on include/fs_utils.h:
```diff
@@ -32,6 +32,37 @@
+
+constexpr int32_t NativeFileSeekFailed = -1;
+
+struct NativeFileReturn
```
`NativeFileIoResult` would be better.

weirddan455 replied:
renamed

johnnovak commented on include/fs_utils.h:
```diff
@@ -32,6 +32,37 @@
+	bool error = false;
+};
+
+enum class NativeFileSeek
```
`NativeSeek`?

weirddan455 replied:
renamed

shermp commented:
Is there a case for making these file i/o wrappers more generic so that other parts of the code could use them later. I'm talking about seek, read, write etc.

Other code that could use this would be the disk image code, in which case large file support would definitely be appreciated.

weirddan455 commented:
> Is there a case for making these file i/o wrappers more generic so that other parts of the code could use them later. I'm talking about seek, read, write etc.
> 
> Other code that could use this would be the disk image code, in which case large file support would definitely be appreciated.

They're already fairly generic and should be fine to use elsewhere.  Some of the calls (especially create and open) may need additional flags depending on the use-case but I think we can cross that bridge when we come to it.

Maybe I should rename the functions to `create_native_file` `read_native_file`, etc.  I'll bump the size types up to `int64_t` as @johnnovak suggested as well.

FeralChild64 reviewed (commented):
Looks OK. I have no remarks.

So far tried some Windows 3.1 software, Colonization game. and MkS_Vir virus scanner, didn't find any problems/

weirddan455 commented:
Thanks for testing @FeralChild64 

I pushed changes addressing the comments above.  Mostly re-naming things and fixing formatting.  I also changed to `int64_t` everywhere in these native I/O functions.

They should be suitable for use where-ever else in the code they might be needed.  I renamed them as I said above to `open_native_file` and such (removing the `local_drive` prefix I had before).  I also made the fat attributes on `create_native_file` a `std::optional` as that one doesn't make sense if we want to create a normal file on the host.

johnnovak commented on src/dos/drive_local.cpp:
```diff
@@ -697,65 +622,54 @@ bool localFile::Write(uint8_t *data, uint16_t *size)
+
+	auto returned_pos = seek_native_file(file_handle, requested_pos, seek_type);
+
+	// TODO: Test Black Throne and see if this hack is still needed
```
Could you test this before merging?

weirddan455 replied:
Tested.  It worked with this hack in and failed to launch with it removed.  I pushed a new commit with more accurate/less hacky behavior and tested this works on Linux and Windows.

I was side-eyeing that `reinterpret_cast` and it turned out to be incorrect for `SEEK_SET`.  The game seeks to 2777296400 (over 2GB) into the file.  Windows and POSIX both allow this.  It doesn't actually resize the file on seek but if you then write to the file, it will fill with zeros up to that size.

Anyway, we just need to interpret the value as unsigned if using `SEEK_SET`.  I put some references to MS-DOS Encyclopedia in the comments.

The error handling and set to 0 isn't needed by Blackthorne but may be needed by other DOS games doing weird things.  Apparently the DOS Seek interrupt never fails assuming you give it a valid handle and seek type.

johnnovak replied:
Thanks for the detailed explanation @weirddan455. This seems to have been quite a rabbit hole, but I'm glad you got to the bottom of it 🎉

weirddan455 replied:
It didn't actually take me too long.  Luckily it was an easily reproducible error and fail to launch on Blackthorne.  Then I just added a log statement and saw it was being cast to a large negative seek value, where it should have always been positive.

johnnovak reviewed (approved):
This is solid step in the right direction @weirddan455 🎉 Looks very good now, left one last comment about testing Blackthorne if you have the capacity. Great work!

weirddan455 commented:
I'll let you take a look at my last commit if you like @johnnovak addressing the Blackthrone issue.

weirddan455 commented:
Added a `LOG_WARNING` if the file seek fails.  I wasn't quite sure what to do in this case because apparently the real DOS interrupt never fails.  The most likely case is a game trying to seek before the beginning of the file so I'm just setting the file position to 0.

Blackthorn never hits this case.  Using unsigned integer lets it succeed.  I wanted this warning to be visible to end users rather than a debug log because it seems to be rare and I want users to put this in a bug report if my assumption proves to be incorrect.

johnnovak commented on src/dos/drive_local.cpp:
```diff
@@ -642,33 +642,45 @@ bool localFile::Seek(uint32_t *pos_addr, uint32_t type)
-	*reinterpret_cast<int32_t*>(pos_addr) = check_cast<int32_t>(returned_pos);
+	// The returned value is always positive.
+	// It can exceed 32-bit signed max (ex. Blackthorne)
+	*pos_addr = check_cast<uint32_t>(returned_pos);
```
Very nice coments, @weirddan455, that's medal-worthy 🎖️

One minor thing, if you'd include the book page numbers next to the citation, that would be helpful. Entirely optional, though, if you're sick of this PR already 😆

weirddan455 replied:
Sure, added the page numbers.

johnnovak replied:
Noice 😎

johnnovak reviewed (approved):
Outstanding work @weirddan455, I'm glad you've managed to remove the Blackthorne hack as well 🚀 I agree with your last note about logging a warning, that's the best we can do.
